docroot/modules/custom/dct_sponsors/dct_sponsors.module

```diff
@@ -23,3 +28,36 @@ function dct_sponsors_theme() {
   ];
 }
 
+/**
+ * Implements hook_ENTITY_TYPE_access().
+ */
+function dct_sponsors_node_access(NodeInterface $node, $op, AccountInterface $account) {
+  // Checks if the node is of type sponsor.
+  if ($node->bundle() == 'sponsor') {
+    $sponsor = \Drupal::entityTypeManager()->getStorage('node')->load($node->id());
+    // The id of the sponsor editor.
+    $user_id = $sponsor->get('field_sponsor_editor')->getValue()[0]['target_id'];
+    if ($user_id) {
+      if ($account->id() == $user_id) {
+        return AccessResult::allowed();
+      }
+    }
+  }
```

Comment:
Node access has to return something, if it can't decide the access, please return access neutral.